### DISCUSSION TITLE
.NET C# library link

### DISCUSSION THREAD

**Discussion by XTOCON** [ERGO-Code/HiGHS#1035] [Q&A] [answered]
I'm using Visual Studio 2022 on Windows to build HiGHS which has succesfully generated highs.dll and highs.exe

Could you please explain how to use these libraries in a .NET C# project?

Thank you

Christophe

**Comment by pjaborges** [comment]
This is actually relevant for me too!
Can you have access to the contents of highs.dll if you include the dll in a C# project?

Paulo

**Comment by pjaborges** [comment]
Hi!

I think should be something around this:
https://stackoverflow.com/questions/40215388/c-sharp-import-dll-function-with-const-string-as-parameter

**Reply by XTOCON** [reply]
Yeah, something around that, but an example of how to call highs library specifically would be very useful.

Do we have a demo project available that calls these libraries in a C# project?

Is there a description of the functions available in the library that we can call?

**Reply by jajhall** [reply]
No-one in the HiGHS team can address this, sorry

**Comment by XTOCON** [comment, marked as answer]
Hi, thank you for your reply, no worries, I'll try to figure it out myself

**Reply by pjaborges** [reply]
Hi! I wish you all the best on that. As I said this is something also highly relevant for me but unfortunately at the moment I don't have the time to dig into it. So I am using the 'dirty' approach by generating the LP format file. This is of course is very inefficient if I would like to add some more constraints or change the objective function to an existing model.

Saying this, It will be great if you could share your findings here! ;) Otherwise I let you know (here) when I will start on this!

**Reply by tosttost** [reply]
In fact calling Highs from C# is quite easy:
* https://github.com/ERGO-Code/HiGHS/blob/master/src/interfaces/highs_csharp_api.cs contains all the PInvoke you need. Copy it into your C# project (check license first,  although MIT is easy to respect).
* Make sure that the native highs library (highs.dll, libhighs.dll, libhighs.so, ... depending on your platform) can be found at runtime. How to do this is platform dependent, copying it next to your C# executable should work in most cases. You can use msbuild for that. At least on linux installing highs system wide should work, too.
* Make sure that all dependencies of the highs library can be found, too. E.g. if Highs was build using Visual C++ make sure that the MSVCRuntime is installed on the machine you want to run your application on.
* depending on the name of your highs library it might be necessary to change the constant "highslibname", see https://learn.microsoft.com/en-us/dotnet/standard/native-interop/cross-platform if necessary. 
* Call the Methods in highs_csharp_api.cs and have fun with HiGHS.

This is the normal way to call plain old C from C# with the great simplification that you don't have to write the PInvoke declarations yourself.

Obviously it would be better to put the C# interface on nuget, ideally with a option for prebuild binaries (although distributing prebuild binaries can be tricky). Maybe one could reuse the binaries from the Julia package. My employer has recently introduced rather controversial rules for making contributions to open source projects. Therefore I'm not sure I'm allowed to do this in an sane way.

**Reply by jannicklange** [reply]
Maybe this is still relevant for you:
We provide a Wrapper for the HiGHS C# Interface in the OPTANO.Modeling NuGet package:
* https://www.nuget.org/packages/OPTANO.Modeling

The package provides an intuitive API for formulating mathematical models, alongside with adapters for the most common MIP solvers (Gurobi, CPLEX, GLPK, HiGHS, ...). Personally, I prefer writing in the modeling API, since it provides many quality-of-life features, and the code can be used with any of the provided adapters, without writing "native" code for all different APIs.

Even though the HiGHS adapter was built against HiGHS 1.3.x, it still seems to be compatible with the latest master release - at least my local tests ran through without issues.
* **UPDATE**
  * The `HighsModelStatus.kSolutionLimit` is not supported by the existing `Modeling.HighsSolver13x` adapter.
  * If your optimization will not return the code, you can proceed with the existing 1.3.x adapter. Otherwise, you'll have to use the 1.4.x adapter, which I will publish shortly on NuGet.

To get started with modeling, you can find some example problems here:
* https://docs.optano.com/modeling/current/userDoc/model_examples/advanced_introduction.html
* https://github.com/OPTANO/optano.modeling.examples

If you have any questions, don't hesitate to contact us:
* https://groups.google.com/u/1/g/optano-modeling

Best Regards,
Jannick

**Reply by jajhall** [reply]
Thanks, I'll add this to the `C#` information

**Reply by ChristopheLauwerys** [reply]
Thank you Jannick for the information, I'll have a look at it and get back to you if I have any more questions
